clippy_lints/src/methods/single_or_empty_collections_iter.rs

```diff
@@ -0,0 +1,201 @@
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::higher::VecArgs;
+use clippy_utils::macros::root_macro_call_first_node;
+use clippy_utils::msrvs::{Msrv, ITER_ONCE_AND_EMPTY};
+use clippy_utils::source::snippet_opt;
+use clippy_utils::{get_parent_expr, is_lang_item_or_ctor, last_path_segment, path_res, std_or_core};
+use rustc_errors::{Applicability, Diagnostic};
+use rustc_hir::{Expr, ExprKind, GenericArg, LangItem};
+use rustc_lint::{LateContext, Lint, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_middle::ty;
+use rustc_span::{sym, Span};
+
+use super::{ITER_ON_EMPTY_COLLECTIONS, ITER_ON_SINGLE_ITEMS};
+
+#[derive(Clone, Copy, Debug)]
+enum Variant<'tcx> {
+    SomeToOnce,
+    NoneToEmpty,
+    OneLenToOnce(&'tcx Expr<'tcx>),
+    ZeroLenToEmpty,
+}
+
+impl<'tcx> Variant<'tcx> {
+    fn as_lint(self) -> &'static Lint {
+        match self {
+            Self::SomeToOnce | Self::OneLenToOnce(_) => ITER_ON_SINGLE_ITEMS,
+            Self::NoneToEmpty | Self::ZeroLenToEmpty => ITER_ON_EMPTY_COLLECTIONS,
+        }
+    }
+
+    fn as_str(self) -> &'static str {
+        match self {
+            Self::SomeToOnce => "Some",
+            Self::NoneToEmpty => "None",
+            Self::OneLenToOnce(_) => "[T; 1]",
+            Self::ZeroLenToEmpty => "[T; 0]",
+        }
+    }
+
+    fn desc(self) -> &'static str {
+        match self {
+            Self::SomeToOnce | Self::OneLenToOnce(_) => "iterator with only one element",
+            Self::NoneToEmpty | Self::ZeroLenToEmpty => "empty iterator",
+        }
+    }
+
+    fn sugg_fn(self, cx: &LateContext<'_>) -> Option<String> {
+        Some(format!(
+            "{}::iter::{}",
+            std_or_core(cx)?,
+            match self {
+                Self::SomeToOnce | Self::OneLenToOnce(_) => "once",
+                Self::NoneToEmpty | Self::ZeroLenToEmpty => "empty",
+            },
+        ))
+    }
```

Comment:
Why does this function return `Option<String>`? Just forgotten to change it or does it have a meaning?

Reply:
`std_or_core(cx)?,`